### PR TITLE
UCP/PERFTEST: Wait for last atomic

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/tools/perf/lib/ucp_tests.cc [hunk 1]
```diff
@@ -418,17 +445,10 @@ class ucp_perf_test_runner {
             /* coverity[switch_selector_expr_is_constant] */
             switch (TYPE) {
             case UCX_PERF_TEST_TYPE_PINGPONG:
-                ptr = (volatile uint8_t*)buffer + length - 1;
-                while (*ptr != sn) {
-                    progress_responder();
-                }
+                wait_for_sn(buffer, sn);
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:448]: seems after change we read buffer head, before change we read buffer tail (+length-1)
pls also see comments on #7013

File: src/tools/perf/lib/ucp_tests.cc [hunk 2]
```diff
@@ -279,6 +279,34 @@ class ucp_perf_test_runner {
...
+                                 &sn, UCS_MEMORY_TYPE_HOST,
+                                 sizeof(sn));
+    }
+
+    void UCS_F_ALWAYS_INLINE wait_for_sn(void *buffer, psn_t sn,
+                                         bool wait_mem = false)
+    {
+        psn_t host_sn;
+
+        while (1) {
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:294]: minor: for (;;)

File: src/tools/perf/lib/ucp_tests.cc (hunk 1 above)
Comment (Artemy-Mellanox) [reply src/tools/perf/lib/ucp_tests.cc:448]: yes. I'm not certain with this tail - AFAIU _length_ is length with contig datatype and iov count with iov.

File: src/tools/perf/lib/libperf.c [hunk 3]
```diff
@@ -1087,6 +1088,10 @@ static ucs_status_t ucp_perf_test_alloc_mem(ucx_perf_context_t *perf)
         buffer_size = ucx_perf_get_message_size(params);
     }
 
+    if (ucx_perf_cmd_is_atomic(params->command)) {
+        buffer_size *= 2;
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:1092]: discussed offline: need to use one buffer, and use the payload to signal the last iteration (e.g put(1), fadd(1), cswap(0 to 1)

File: src/tools/perf/lib/libperf_int.h [hunk 4]
```diff
@@ -231,6 +231,15 @@ size_t ucx_perf_get_message_size(const ucx_perf_params_t *params)
     return length;
 }
 
+static inline int ucx_perf_cmd_is_atomic(ucx_perf_cmd_t command) {
+    if ((command == UCX_PERF_CMD_ADD) || (command == UCX_PERF_CMD_FADD) ||
+        (command == UCX_PERF_CMD_SWAP) || (command == UCX_PERF_CMD_CSWAP)) {
+        return 1;
+    }
+
+    return 0;
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:240]: return (... ) || ( )

File: src/tools/perf/lib/ucp_tests.cc [hunk 5]
```diff
@@ -466,25 +466,54 @@ class ucp_perf_test_runner {
...
      * this message */
-    ucs_status_t send_last_iter(ucp_ep_h ep, void *buffer,
+    ucs_status_t send_last_iter(ucp_ep_h ep, void *buffer, size_t size,
                                 uint64_t remote_addr, ucp_rkey_h rkey)
     {
-        if ((CMD == UCX_PERF_CMD_PUT) &&
-            (TYPE == UCX_PERF_TEST_TYPE_STREAM_UNI)) {
+        ucs_status_ptr_t status_p;
+
+        if (use_psn()) {
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:489]: if (!use_psn) return OK;

File: test/gtest/common/test_perf.h [hunk 6]
```diff
@@ -16,11 +16,13 @@ class test_perf {
 protected:
     struct test_spec {
         const char             *title;
+        const char             *short_title;
```
Comment (yosefe) [inline test/gtest/common/test_perf.h:19]: maybe keep only one title (short)?

File: test/gtest/ucp/test_ucp_perf.cc [hunk 7]
```diff
@@ -17,8 +18,34 @@
...
 class test_ucp_perf : public ucp_test, public test_perf {
 public:
+    enum {
+        TEST_TYPE,
+        ATOMIC_MODE
+    };
+
+    enum {
+        CPU = 1,
+        DEVICE
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_perf.cc:28]: ATOMIC_CPU

File: test/gtest/ucp/test_ucp_perf.cc [hunk 8]
```diff
@@ -17,8 +18,34 @@
 
 class test_ucp_perf : public ucp_test, public test_perf {
 public:
+    enum {
+        TEST_TYPE,
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_perf.cc:22]: VARIANT_TEST_TYPE

File: test/gtest/ucp/test_ucp_perf.cc [hunk 9]
```diff
@@ -17,8 +18,34 @@
...
+        DEVICE
+    };
+
     static void get_test_variants(std::vector<ucp_test_variant>& variants) {
-        add_variant(variants, 0);
+        ucp_test_variant* variant;
+        int i = 0;
+
+        for (const test_spec *test = tests; test->title != NULL; ++test, ++i) {
+            if (test->ctx_features == UCP_FEATURE_AMO) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_perf.cc:36]: maybe check perf_cmd, to not add new field ctx_features?

File: test/gtest/ucp/test_ucp_perf.cc [hunk 10]
```diff
@@ -17,8 +18,34 @@
...
+        CPU = 1,
+        DEVICE
+    };
+
     static void get_test_variants(std::vector<ucp_test_variant>& variants) {
-        add_variant(variants, 0);
+        ucp_test_variant* variant;
+        int i = 0;
+
+        for (const test_spec *test = tests; test->title != NULL; ++test, ++i) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_perf.cc:35]: maybe do for (i=..) and then test=&tests[i]

File: src/tools/perf/lib/ucp_tests.cc [hunk 11]
```diff
@@ -466,25 +466,54 @@ class ucp_perf_test_runner {
...
             fence();
             *(uint8_t*)buffer = UCP_PERF_LAST_ITER_SN;
-            return ucp_put(ep, buffer, sizeof(uint8_t), remote_addr, rkey);
+            /* coverity[switch_selector_expr_is_constant] */
+            switch (CMD) {
+            case UCX_PERF_CMD_PUT:
+                return ucp_put(ep, buffer, sizeof(uint8_t), remote_addr, rkey);
```
Comment (yosefe) [inline src/tools/perf/lib/ucp_tests.cc:495]: seems return value is unused
maybe this func should wait for request completion?

Comment (Artemy-Mellanox) [reply src/tools/perf/lib/ucp_tests.cc:495]: it does this operation and than flush().
shouldn't flush wait for requests to complete?

Comment (yosefe) [reply src/tools/perf/lib/ucp_tests.cc:495]: yes but also need to release the request

Comment (Artemy-Mellanox) [reply src/tools/perf/lib/ucp_tests.cc:495]: do it from nop_cb, no?

Comment (yosefe) [reply src/tools/perf/lib/ucp_tests.cc:495]: ok

**Review by yosefe — APPROVED** [review]

**Comment by Artemy-Mellanox** [comment]
bot:pipe:retest

**Pull request merged**